vdc/scripts/users/user_data.py

```diff
@@ -0,0 +1,155 @@
+from table import Table
+
+shortuuid.set_alphabet("0123456789abcdefghijkmnopqrstuvwxyz")
+
+
+def create_service_id():
+    return f'user-{shortuuid.uuid()}'
+
+
+def create_google_service_account(sa_name, google_project):
+    return gcloud_service.projects().serviceAccounts().create(
+            name=f'projects/{google_project}',
+            body={
+                "accountId": sa_name, "serviceAccount": {
+                    "displayName": "user"
+                }
+            }).execute()
+
+
+def delete_google_service_account(gsa_email, google_project):
+    return gcloud_service.projects().serviceAccounts().delete(name=f'projects/-/serviceAccounts/{gsa_email}').execute()
+
+
+def create_kube_service_acccount(namespace):
+    return k8s.create_namespaced_service_account(
+        namespace=namespace,
+        body=kube_client.V1ServiceAccount(
+            api_version='v1',
+            metadata=kube_client.V1ObjectMeta(
+                generate_name='user-',
+                annotations={
+                    "type": "user"
+                }
+            )
+        )
+    )
+
+
+def delete_kube_service_acccount(ksa_name, namespace):
+    return k8s.delete_namespaced_service_account(name=ksa_name, namespace=namespace, body={})
+
+
+def create_bucket(sa_name, gsa_email):
+    bucket = storage.Client().bucket(sa_name)
+    bucket.labels = {
+        'type': 'user',
+    }
+
+    bucket.create()
+
+    acl = bucket.acl
+    acl.user(gsa_email).grant_owner()
+
+    return bucket
+
+
+def delete_bucket(bucket_name):
+    return storage.Client().get_bucket(bucket_name).delete()
+
+
+def create_all(google_project, kube_namespace):
+    out = {}
+
+    ksa_response = create_kube_service_acccount(kube_namespace)
+    out['ksa_name'] = ksa_response.metadata.name
+
+    sa_name = create_service_id()
+
+    gs_response = create_google_service_account(sa_name, google_project)
+    out['gsa_email'] = gs_response['email']
+
+    create_bucket(sa_name, out['gsa_email'])
+    out['bucket_name'] = sa_name
+
+    return out
+
+
+def delete_all(user_obj, google_project='hail-vdc', kube_namespace='default'):
+    modified = 0
+    try:
+        delete_bucket(user_obj['bucket_name'])
+        modified += 1
+    except google.api_core.exceptions.NotFound:
+        pass
+
+    try:
+        delete_google_service_account(user_obj['gsa_email'], google_project)
+        modified += 1
+    except HttpError as e:
+        if e.resp.status != 404:
+            raise e
+
+    try:
+        delete_kube_service_acccount(user_obj['ksa_name'], kube_namespace)
+        modified += 1
+    except kube_client.rest.ApiException as e:
+        if e.status != 404:
+            raise e
+
+    if modified == 0:
+        return 404
+
+
+def create_all_idempotent(user_id, google_project='hail-vdc', kube_namespace='default'):
+    table = Table()
+    existing = table.get(user_id)
+
+    if existing is None:
+        res = create_all(google_project, kube_namespace)
```

Comment:
How is this idempotent?

Reply:
Because if the resources were already created, the user doesn't get any additional resources. So calling the function 1 + N times gives the same result

Reply:
ok. got it.